### PR TITLE
Handling attributes, ranges and their changes in (Live)Selection.

### LINKED ISSUE #259: Check whether it's optimal that selection updates its own attributes on #changesDone and #change:range
Quote from https://github.com/ckeditor/ckeditor5-core/pull/234

> I'm not sure on what events the selection should update its attributes itself. This should not conflict with the situation when the feature sets the selection and its attributes (e.g. bold command when executed will add attributes to the selection and that will, if I understand the flow correctly, trigger #changesDone).

### PR DESCRIPTION
**Pull request by scofalik**
Fixes #259. See the issue for the list of changes.

Followup issue: #633

Related PR: https://github.com/ckeditor/ckeditor5-core/pull/31

### CODE REVIEW

File: src/model/schema.js [hunk 1]
```diff
@@ -178,6 +178,11 @@ export class SchemaItem {
 
 			// Now we have to check every item name from the path to check.
 			for ( let checkName of checkPath ) {
+				// Don't check items that are not registered in schema.
+				if ( !this._schema.hasItem( checkName ) ) {
```
Comment (Reinmar) [inline src/model/schema.js]: Missing test for this change.

**Comment by Reinmar** [comment]
I found one regression: https://github.com/ckeditor/ckeditor5-engine/issues/637.

**Comment by Reinmar** [comment]
The rest is fine – 100% CC, manual tests pass, etc. #637 needs to be changed in other piece of code, so it's independent.

**Comment by Reinmar** [comment]
The other tickets which will be closed or needs to be checked:

* https://github.com/ckeditor/ckeditor5-engine/issues/589
* https://github.com/ckeditor/ckeditor5-engine/issues/603 (needs to be checked)
* https://github.com/ckeditor/ckeditor5-engine/issues/462

**Comment by Reinmar** [comment]
I pushed manual test for #603 and checked that it's fully fixed by this PR. Cool!

File: src/model/liveselection.js [hunk 2]
```diff
@@ -42,10 +42,41 @@ export default class LiveSelection extends Selection {
...
+		 * @protected
+		 * @member {engine.model.Document} engine.model.LiveSelection#_document
 		 */
 		this._document = document;
+
+		/**
+		 * Keeps mapping of attribute name to priority with which the attribute got modified (added/changed/removed)
+		 * last time. Possible values of priority are: `'low'` and `'normal'`.
+		 *
+		 * Priorities are used by internal `LiveSelection` mechanisms. All attributes set using `LiveSelection`
```
Comment (scofalik) [inline src/model/liveselection.js]: `low` is being used when attributes are changed indirectly, that is through `_updateAttributes`.

Really, if you think is is overengineered I'll try to close this "logic" (that `_updateAttributes` clears only attributes set by itself) in `_updateAttributes`. IMHO it is over-engineered, but we may resolve this issue in ... spearate issue.

File: src/model/liveselection.js [hunk 3]
```diff
@@ -42,10 +42,41 @@ export default class LiveSelection extends Selection {
...
+		 * @private
+		 * @member {Map} engine.model.LiveSelection#_attributePriority
+		 */
+		this._attributePriority = new Map();
+
+		// Whenever selection range changes, if the change comes directly from selection API (direct user change).
+		this.on( 'change:range', ( evt, data ) => {
+			if ( data.directChange ) {
+				// Reset attributes on selection (clear attributes and priorities).
+				//this._resetAttributes();
```
Comment (scofalik) [inline src/model/liveselection.js]: Te commented code is a mistake. Sorry for that.

As far as `_updateAttributes` is concerned, this issue may be resolved by simplifying `LiveSelection` internal logic and getting rid of priorities. I thought priorities, as a general solution, will be nice, but they got too clunky and even though I have those priorities, I still have to write some special code (like in `_updateAttributes`).

**Comment by scofalik** [comment]
I've pushed test for `Schema` and removed commented line of code.
As I understand #603 is fine white #589 and #462 are to-be-closed.
#637 is a different case and will be handled separately.

Does this mean that this PR is ready to be closed?

**Comment by Reinmar** [comment]
Another ticket which seemed to be fixed: https://github.com/ckeditor/ckeditor5-engine/issues/590.

**Pull request merged**